File: autosave.py
"""自动保存和编辑日志工作进程。

服务端把磁盘写入工作交给这个模块中的 worker 进程完成。这样主服务端线程
处理网络请求和文档锁时，不会因为文件系统 I/O 变慢而卡住其他客户端。
"""

from __future__ import annotations

import json
import signal
import shutil
import time
from multiprocessing import Queue
from pathlib import Path
from queue import Empty
from typing import Any, Dict, Optional


def atomic_write_text(path: Path, content: str) -> None:
    # 先写到临时文件，再用 replace 一次性替换目标文件。
    # 这样即使程序在写入中途崩溃，也尽量避免留下半截正式文件。
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp_path = path.with_name(path.name + ".tmp")
    tmp_path.write_text(content, encoding="utf-8")
    tmp_path.replace(path)


def write_with_backup(path: Path, content: str) -> None:
    # 手动保存会覆盖正式文件，所以覆盖前先复制一份 .bak。
    # 自动保存只写 .autosave，不走这里。
    path.parent.mkdir(parents=True, exist_ok=True)
    if path.exists():
        backup_path = path.with_name(path.name + ".bak")
        shutil.copy2(path, backup_path)
    atomic_write_text(path, content)


def append_json_log(path: Path, entry: Dict[str, Any]) -> None:
    # 日志使用 JSON Lines：每一行是一个独立 JSON 对象，便于 tail 查看，
    # 也便于后续用脚本逐行分析。
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("a", encoding="utf-8") as handle:
        handle.write(json.dumps(entry, ensure_ascii=False, sort_keys=True) + "\n")


def autosave_worker(
    event_queue: Queue,
    document_path: str,
    log_path: str,
    interval_seconds: float = 5.0,
) -> None:
    # 保存进程不响应 Ctrl-C，由父进程负责发送 STOP 做有序退出。
    signal.signal(signal.SIGINT, signal.SIG_IGN)
    # 自动保存和日志写入放在单独进程中完成。
    # 主服务端只需要通过 multiprocessing.Queue 投递事件，不必在处理
    # 网络请求或文档锁时直接做磁盘 I/O，从而体现进程间通信的使用。
    target_path = Path(document_path)
    autosave_path = target_path.with_name(target_path.name + ".autosave")
    log_file = Path(log_path)
    latest_content: Optional[str] = None
    latest_version = 0
    # 使用 monotonic 计算间隔，避免系统时间被手动调整影响自动保存节奏。
    next_save = time.monotonic() + interval_seconds

    while True:
        # 队列 get 的超时时间正好是“距离下一次自动保存还剩多久”。
        # 如果超时没有新事件，就执行一次定时保存检查。
        timeout = max(0.0, next_save - time.monotonic())
        try:
            event = event_queue.get(timeout=timeout)
        except Empty:
            event = None

        if event:
            event_type = event.get("type")
            if event_type == "STOP":
                # 退出前把最后一份快照写入 autosave 文件，减少关闭时的数据丢失窗口。
                if latest_content is not None:
                    atomic_write_text(autosave_path, latest_content)
                return
            if event_type == "SNAPSHOT":
                # 服务端每处理完一次成功编辑，就发送最新快照。
                # 工作进程只保存最后一次快照，定时落盘，避免每个按键都写文件。
                latest_content = event.get("content", "")
                latest_version = int(event.get("version", latest_version))
            elif event_type == "SAVE_NOW":
                # 手动保存写入正式文件，并在覆盖前生成 .bak 备份。
                content = event.get("content", latest_content or "")
                latest_content = content
                latest_version = int(event.get("version", latest_version))
                write_with_backup(target_path, content)
                append_json_log(
                    log_file,
                    {
                        "event": "manual_save",
                        "version": latest_version,
                        "path": str(target_path),
                        "timestamp": time.time(),
                    },
                )
            elif event_type == "LOG":
                # 普通编辑日志不改文档文件，只追加一行 JSON。
                entry = dict(event.get("entry", {}))
                entry.setdefault("timestamp", time.time())
                append_json_log(log_file, entry)

        if time.monotonic() >= next_save:
            if latest_content is not None:
                # 自动保存写的是旁路快照文件，不覆盖正式文件。
                atomic_write_text(autosave_path, latest_content)
                append_json_log(
                    log_file,
                    {
                        "event": "autosave",
                        "version": latest_version,
                        "path": str(autosave_path),
                        "timestamp": time.time(),
                    },
                )
            next_save = time.monotonic() + interval_seconds

File: autosave.c
#include "termsync.h"

#include <errno.h>
#include <fcntl.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/select.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <time.h>
#include <unistd.h>

static int mkdir_parent_recursive(const char *path)
{
    char copy[PATH_MAX];
    snprintf(copy, sizeof(copy), "%s", path);
    char *slash = strrchr(copy, '/');
    if (slash == NULL) {
        return 0;
    }
    *slash = '\0';
    if (copy[0] == '\0') {
        return 0;
    }
    for (char *p = copy + 1; *p != '\0'; ++p) {
        if (*p == '/') {
            *p = '\0';
            if (mkdir(copy, 0775) != 0 && errno != EEXIST) {
                return -1;
            }
            *p = '/';
        }
    }
    if (mkdir(copy, 0775) != 0 && errno != EEXIST) {
        return -1;
    }
    return 0;
}

static void suffix_path(const char *path, const char *suffix, char *out, size_t out_size)
{
    snprintf(out, out_size, "%s%s", path, suffix);
}

static int copy_file(const char *src, const char *dst)
{
    FILE *in = fopen(src, "rb");
    if (in == NULL) {
        return -1;
    }
    FILE *out = fopen(dst, "wb");
    if (out == NULL) {
        fclose(in);
        return -1;
    }
    char buffer[8192];
    size_t n;
    while ((n = fread(buffer, 1, sizeof(buffer), in)) > 0) {
        if (fwrite(buffer, 1, n, out) != n) {
            fclose(in);
            fclose(out);
            return -1;
        }
    }
    fclose(in);
    fclose(out);
    return 0;
}

int ts_atomic_write_text(const char *path, const char *content)
{
    char tmp_path[PATH_MAX];
    if (mkdir_parent_recursive(path) != 0) {
        return -1;
    }
    suffix_path(path, ".tmp", tmp_path, sizeof(tmp_path));
    FILE *fp = fopen(tmp_path, "wb");
    if (fp == NULL) {
        return -1;
    }
    const char *text = content ? content : "";
    size_t len = strlen(text);
    if (fwrite(text, 1, len, fp) != len) {
        fclose(fp);
        return -1;
    }
    if (fclose(fp) != 0) {
        return -1;
    }
    return rename(tmp_path, path);
}

int ts_write_with_backup(const char *path, const char *content)
{
    char backup_path[PATH_MAX];
    if (mkdir_parent_recursive(path) != 0) {
        return -1;
    }
    if (access(path, F_OK) == 0) {
        suffix_path(path, ".bak", backup_path, sizeof(backup_path));
        if (copy_file(path, backup_path) != 0) {
            return -1;
        }
    }
    return ts_atomic_write_text(path, content);
}

int ts_append_json_log(const char *path, const char *entry_json)
{
    if (mkdir_parent_recursive(path) != 0) {
        return -1;
    }
    FILE *fp = fopen(path, "ab");
    if (fp == NULL) {
        return -1;
    }
    fprintf(fp, "%s\n", entry_json ? entry_json : "{}");
    return fclose(fp);
}

static long monotonic_ms(void)
{
    struct timespec now;
    clock_gettime(CLOCK_MONOTONIC, &now);
    return now.tv_sec * 1000L + now.tv_nsec / 1000000L;
}

static int read_line_fd(int fd, char *line, size_t line_size)
{
    size_t len = 0;
    while (len + 1 < line_size) {
        char ch;
        ssize_t n = read(fd, &ch, 1);
        if (n < 0) {
            if (errno == EINTR) {
                continue;
            }
            return -1;
        }
        if (n == 0) {
            return 0;
        }
        if (ch == '\n') {
            line[len] = '\0';
            return 1;
        }
        line[len++] = ch;
    }
    line[len] = '\0';
    return -1;
}

static char *read_exact_alloc(int fd, size_t len)
{
    char *data = malloc(len + 1);
    if (data == NULL) {
        return NULL;
    }
    size_t used = 0;
    while (used < len) {
        ssize_t n = read(fd, data + used, len - used);
        if (n < 0) {
            if (errno == EINTR) {
                continue;
            }
            free(data);
            return NULL;
        }
        if (n == 0) {
            free(data);
            return NULL;
        }
        used += (size_t)n;
    }
    data[len] = '\0';
    return data;
}

static int send_payload_event(int fd, const char *name, const char *content, int version)
{
    const char *payload = content ? content : "";
    char header[128];
    int header_len = snprintf(header, sizeof(header), "%s %d %zu\n", name, version, strlen(payload));
    if (header_len < 0) {
        return -1;
    }
    if (ts_write_all_fd(fd, header, (size_t)header_len) != 0) {
        return -1;
    }
    return ts_write_all_fd(fd, payload, strlen(payload));
}

int ts_autosave_send_snapshot(int fd, const char *content, int version)
{
    return send_payload_event(fd, "SNAPSHOT", content, version);
}

int ts_autosave_send_save_now(int fd, const char *content, int version)
{
    return send_payload_event(fd, "SAVE_NOW", content, version);
}

int ts_autosave_send_log(int fd, const char *entry_json)
{
    const char *payload = entry_json ? entry_json : "{}";
    char header[128];
    int header_len = snprintf(header, sizeof(header), "LOG 0 %zu\n", strlen(payload));
    if (header_len < 0) {
        return -1;
    }
    if (ts_write_all_fd(fd, header, (size_t)header_len) != 0) {
        return -1;
    }
    return ts_write_all_fd(fd, payload, strlen(payload));
}

int ts_autosave_send_stop(int fd)
{
    return ts_write_all_fd(fd, "STOP\n", 5);
}

static void autosave_loop(int fd, const char *document_path, const char *log_path, long interval_ms)
{
    char autosave_path[PATH_MAX];
    suffix_path(document_path, ".autosave", autosave_path, sizeof(autosave_path));
    char *latest_content = NULL;
    int latest_version = 0;
    long next_save = monotonic_ms() + interval_ms;

    while (true) {
        long now = monotonic_ms();
        long wait_ms = next_save > now ? next_save - now : 0;
        fd_set set;
        FD_ZERO(&set);
        FD_SET(fd, &set);
        struct timeval tv;
        tv.tv_sec = wait_ms / 1000;
        tv.tv_usec = (wait_ms % 1000) * 1000;
        int ready = select(fd + 1, &set, NULL, NULL, &tv);
        if (ready > 0 && FD_ISSET(fd, &set)) {
            char header[128];
            int rc = read_line_fd(fd, header, sizeof(header));
            if (rc <= 0) {
                break;
            }
            if (strcmp(header, "STOP") == 0) {
                if (latest_content != NULL) {
                    ts_atomic_write_text(autosave_path, latest_content);
                }
                break;
            }
            char event[32];
            int version = 0;
            size_t len = 0;
            if (sscanf(header, "%31s %d %zu", event, &version, &len) != 3) {
                continue;
            }
            char *payload = read_exact_alloc(fd, len);
            if (payload == NULL) {
                break;
            }
            if (strcmp(event, "SNAPSHOT") == 0) {
                free(latest_content);
                latest_content = payload;
                latest_version = version;
                payload = NULL;
            } else if (strcmp(event, "SAVE_NOW") == 0) {
                free(latest_content);
                latest_content = ts_strdup(payload);
                latest_version = version;
                ts_write_with_backup(document_path, payload);
                char entry[PATH_MAX + 160];
                snprintf(entry, sizeof(entry), "{\"event\":\"manual_save\",\"version\":%d,\"path\":\"%s\"}", version, document_path);
                ts_append_json_log(log_path, entry);
            } else if (strcmp(event, "LOG") == 0) {
                ts_append_json_log(log_path, payload);
            }
            free(payload);
        }

        if (monotonic_ms() >= next_save) {
            if (latest_content != NULL) {
                ts_atomic_write_text(autosave_path, latest_content);
                char entry[PATH_MAX + 160];
                snprintf(entry, sizeof(entry), "{\"event\":\"autosave\",\"version\":%d,\"path\":\"%s\"}", latest_version, autosave_path);
                ts_append_json_log(log_path, entry);
            }
            next_save = monotonic_ms() + interval_ms;
        }
    }
    free(latest_content);
}

int ts_start_autosave_process(
    TsAutosaveProcess *process,
    const char *document_path,
    const char *log_path,
    long interval_ms
)
{
    int fds[2];
    if (pipe(fds) != 0) {
        return -1;
    }
    pid_t pid = fork();
    if (pid < 0) {
        close(fds[0]);
        close(fds[1]);
        return -1;
    }
    if (pid == 0) {
        close(fds[1]);
        autosave_loop(fds[0], document_path, log_path, interval_ms <= 0 ? 5000 : interval_ms);
        close(fds[0]);
        _exit(0);
    }
    close(fds[0]);
    process->pid = pid;
    process->write_fd = fds[1];
    return 0;
}

void ts_stop_autosave_process(TsAutosaveProcess *process)
{
    if (process == NULL || process->write_fd < 0) {
        return;
    }
    ts_autosave_send_stop(process->write_fd);
    close(process->write_fd);
    process->write_fd = -1;
    if (process->pid > 0) {
        waitpid(process->pid, NULL, 0);
        process->pid = -1;
    }
}

